test/built-ins/String/prototype/split/this-value-tostring-error.js

```diff
@@ -0,0 +1,45 @@
+// Copyright (C) 2020 Richard Gibson. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-string.prototype.split
+description: Abrupt completion from ToString on the "this" value
+info: |
+  1. Let _O_ be ? RequireObjectCoercible(*this* value).
+  1. If _separator_ is neither *undefined* nor *null*, then
+    1. Let _splitter_ be ? GetMethod(_separator_, @@split).
+    1. If _splitter_ is not *undefined*, then
+      1. Return ? Call(_splitter_, _separator_, &laquo; _O_, _limit_ &raquo;).
+  1. Let _S_ be ? ToString(_O_).
+---*/
+
+function ExpectedError(message) {
+  this.message = message || "";
+}
+ExpectedError.prototype.toString = function () {
+  return "ExpectedError: " + this.message;
+};
+
+var split = String.prototype.split;
+
+var nonStringableReceiver = {};
+nonStringableReceiver.toString = function() { throw new ExpectedError("receiver.toString"); };
```

Comment:
```suggestion
var receiverToString = 0;
nonStringableReceiver.toString = function() {
  receiverToString += 1;
  throw new ExpectedError("receiver.toString");
};
```

and after the `split.call(nonStringableReceiver, splitter` line:

```js
assert.sameValue(receiverToString, 0, 'ToString should not be called on the receiver when the separator has a @@split method.');
```

and at the end, after the last assert.throws, you can check `receiverToString` is 1, without errors being swallowed or re-checked.

Reply:
I'm not sure what this suggestion is getting at. The way I have things now, the only source of an ExpectedError is from ToString on the receiver, which _must not_ take place as part of `split.call(nonStringableReceiver, splitter, Symbol())` (because the Symbol.split path is special) and _must_ take place as part of `split.call(nonStringableReceiver, nonStringableSeparator, Symbol())` (which is inside the `assert.throws` call that looks for ExpectedError). So the presence of a thrown ExpectedError is already a reliable means of determining that `toString` was called and absence of such an error is already a reliable means of determining that it was not, leaving no apparent reason for an invocation counter.